Sources/SwiftMemcache/MemcachedConnection.swift

```diff
@@ -0,0 +1,170 @@
+            /// The continuation for the request stream.
+            requestContinuation: AsyncStream<StreamElement>.Continuation
+        )
+        case finished
+    }
+
+    /// Enum representing the possible errors that can be encountered in `MemcachedConnection`.
+    enum MemcachedConnectionError: Error {
+        /// Indicates that the connection to the server has finished.
+        case connectionFinished
+        /// Indicates that the request has been dropped because the buffer was full.
+        case requestDropped
+        /// Indicates that the request was not enqueued because the stream was in a terminal state.
+        case requestStreamTerminated
+    }
+
+    private var connectionState: ConnectionState
+
+    /// Initialize a new MemcachedConnection.
+    ///
+    /// - Parameters:
+    ///   - host: The host address of the Memcache server.
+    ///   - port: The port number of the Memcache server.
+    ///   - eventLoopGroup: The event loop group to use for this connection.
+    public init(host: String, port: Int, eventLoopGroup: EventLoopGroup) async throws {
+        self.eventLoopGroup = eventLoopGroup
+        let bootstrap = ClientBootstrap(group: self.eventLoopGroup)
+            .channelInitializer { channel in
+                return channel.pipeline.addHandlers([MessageToByteHandler(MemcachedRequestEncoder()), ByteToMessageHandler(MemcachedResponseDecoder())])
+            }
+
+        let rawChannel = try await bootstrap.connect(host: host, port: port).get()
+
+        self.channel = try await eventLoopGroup.next().submit { () -> NIOAsyncChannel<MemcachedResponse, MemcachedRequest> in
+            return try NIOAsyncChannel<MemcachedResponse, MemcachedRequest>(synchronouslyWrapping: rawChannel)
+        }.get()
+
+        let bufferAllocator = ByteBufferAllocator()
+        let (stream, continuation) = AsyncStream<StreamElement>.makeStream()
+        self.connectionState = .running(
+            bufferAllocator: bufferAllocator,
+            requestStream: stream,
+            requestContinuation: continuation
+        )
+    }
+
+    /// Start consuming the requestStream
+    ///
+    /// This function starts consuming the requests from the `requestStream`,
+    /// sending each request to the Memcache server and handling the server's responses.
+    public func run() async {
+        var iterator = self.channel.inboundStream.makeAsyncIterator()
+        switch self.connectionState {
+        case .running(_, let requestStream, let requestContinuation):
+            for await (request, continuation) in requestStream {
+                do {
+                    try await self.channel.outboundWriter.write(request)
+                    let responseBuffer = try await iterator.next()
+
+                    if let response = responseBuffer {
+                        continuation.resume(returning: response)
+                    }
+                } catch {
+                    self.connectionState = .finished
```

Comment:
Just for safety we should switch over the state here again and make sure that we are in the `running` state still.